src/java.base/share/classes/sun/nio/ch/NativeDispatcher.java

```diff
@@ -69,11 +72,28 @@ abstract long writev(FileDescriptor fd, long address, int len)
 
     abstract void close(FileDescriptor fd) throws IOException;
 
-    // Prepare the given fd for closing by duping it to a known internal fd
-    // that's already closed.  This is necessary on some operating systems
-    // (Solaris and Linux) to prevent fd recycling.
-    //
-    void preClose(FileDescriptor fd) throws IOException {
+    /**
+     * Prepare the given file descriptor for closing. If a virtual thread is blocked
+     * on the file descriptor then it is unparked so that it stops polling. On Unix systems,
+     * if a platform thread is blocked on the file descriptor then the file descriptor is
+     * dup'ed to a special fd and the thread signalled so that the syscall fails with EINTR.
+     */
+    final void preClose(FileDescriptor fd, long reader, long writer) throws IOException {
+        if (NativeThread.isVirtualThread(reader) || NativeThread.isVirtualThread(writer)) {
```

Comment:
There used to be a `if (reader != 0 || writer != 0) {...}` check before doing any of this pre-close work, in `DatagramChannelImpl`. Is that no longer needed?

Reply:
> There used to be a if (reader != 0 || writer != 0) {...} check before doing any of this pre-close work, in DatagramChannelImpl. Is that no longer needed?

Looking at the implementation of `NativeThread.isNativeThread()` and `NativeThread.isVirtualThread()`, it already implicitly takes into account the `!= 0` check, so I guess the previous explicit checks before calling those methods was redundant.

Reply:
Right, there are a several places where check for != 0 would be redundant.